scvi/autotune/_autotune.py

```diff
@@ -0,0 +1,346 @@
+        categorical_covariates = [[covariate] for covariate in categorical_covariates]
+        if test_effect_covariates:
+            continious_covariates.append(None)
+            categorical_covariates.append(None)
+
+        self.adata = adata
+        self.model = model
+        self.training_metrics = training_metrics
+        self.metric_functions = metric_functions
+        self.model_hyperparams = model_hyperparams
+        self.trainer_hyperparams = trainer_hyperparams
+        self.plan_hyperparams = plan_hyperparams
+        self.continious_covariates = {
+            "continious_covariates": tune.choice(continious_covariates)
+            if continious_covariates
+            else None
+        }
+        self.categorical_covariates = {
+            "categorical_covariates": tune.choice(categorical_covariates)
+            if categorical_covariates
+            else None
+        }
+
+        if test_effect_hvg:
+            self.test_effect_hvg = {"subset_adata": tune.choice([True, False])}
+        else:
+            self.test_effect_hvg = {"subset_adata": False}
+
+        self.top_hvg = {
+            "top_n": tune.choice(top_hvg) if test_effect_hvg or top_hvg else None
+        }
+
+        self.metrics = training_metrics
+        self.reporter = CLIReporter(
+            metric_columns=training_metrics + list(self.metric_functions.keys())
+        )
+        self.config = {}
+        for att in [
+            self.model_hyperparams,
+            self.trainer_hyperparams,
+            self.plan_hyperparams,
+            self.continious_covariates,
+            self.categorical_covariates,
+            self.test_effect_hvg,
+            self.top_hvg,
+        ]:
+            if att is not None:
+                self.config.update(att)
+        self.batch_key = batch_key
+        self.num_epochs = num_epochs
+
+    def _trainable(self, config):
+        model_config = {}
+        trainer_config = {}
+        plan_config = {}
+        continious_covariate_config = {}
+        categorical_covariate_config = {}
+        hvg_config = {}
+
+        for key in config:
+            if key in self.model_hyperparams:
+                model_config[key] = config[key]
+            elif key in self.trainer_hyperparams:
+                trainer_config[key] = config[key]
+            elif key in self.plan_hyperparams:
+                plan_config[key] = config[key]
+            elif key in self.continious_covariates:
+                continious_covariate_config[key] = config[key]
+            elif key in self.categorical_covariates:
+                categorical_covariate_config[key] = config[key]
+            elif (
+                (self.test_effect_hvg and self.top_hvg) is not None
+                and key in self.test_effect_hvg
+                or self.top_hvg
+            ):
+                hvg_config[key] = config[key]
+
+        if "subset_adata" in hvg_config is True:
```

Comment:
I would split up the above and below into separate functions for readability purposes, the above being formatting the config dictionaries, and the below actually running the model with the respective configuration.